File: Neato-SDK/neato-sdk-android/src/test/java/com/neatorobotics/sdk/android/authentication/NeatoAuthenticationResponseTest.java
package com.neatorobotics.sdk.android.authentication;

import android.net.Uri;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.runners.MockitoJUnitRunner;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertFalse;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class NeatoAuthenticationResponseTest {

    @Mock
    Uri uriToken;
    @Mock
    Uri uriError;
    @Mock
    Uri uriInvalid;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        when(uriToken.toString()).thenReturn("my-app://neato#access_token=123&expiration=1");
        when(uriError.toString()).thenReturn("my-app://neato#error=error_1&error_description=description1");
        when(uriInvalid.toString()).thenReturn("");
    }

    @Test
    public void fromUriToken() throws Exception {
        NeatoAuthenticationResponse response = NeatoAuthenticationResponse.fromUri(uriToken);
        assertNotNull(response);
        assertTrue(response.getType() == NeatoAuthenticationResponse.Response.TOKEN);
        assertEquals("123", response.getToken());
        assertNull(response.getError());
        assertNull(response.getErrorDescription());
    }

    @Test
    public void fromUriError() throws Exception {
        NeatoAuthenticationResponse response = NeatoAuthenticationResponse.fromUri(uriError);
        assertNotNull(response);
        assertTrue(response.getType() == NeatoAuthenticationResponse.Response.ERROR);
        assertEquals("error_1", response.getError());
        assertEquals("description1", response.getErrorDescription());
        assertNull(response.getToken());
    }

    @Test
    public void fromUriCancelled() throws Exception {
        NeatoAuthenticationResponse response = NeatoAuthenticationResponse.fromUri(uriInvalid);
        assertNotNull(response);
        assertTrue(response.getType() == NeatoAuthenticationResponse.Response.INVALID);
        assertNull(response.getError());
        assertNull(response.getErrorDescription());
        assertNull(response.getToken());
    }

    @Test
    public void splitQueryMultipleParams() throws Exception {
        Map<String, String> params = NeatoAuthenticationResponse.fromUri(null).splitQuery("a=vala&b=valb");
        assertNotNull(params);
        assertNotNull(params.get("a"));
        assertNotNull(params.get("b"));
        assertEquals(params.get("a"), "vala");
        assertEquals(params.get("b"), "valb");
    }

    @Test
    public void splitQuerySingleParam() throws Exception {
        Map<String, String> params = NeatoAuthenticationResponse.fromUri(null).splitQuery("a=vala");
        assertNotNull(params);
        assertNotNull(params.get("a"));
        assertNull(params.get("b"));
        assertEquals(params.get("a"), "vala");
    }

    @Test
    public void splitQueryNoParam() throws Exception {
        Map<String, String> params = NeatoAuthenticationResponse.fromUri(null).splitQuery("");
        assertNotNull(params);
        assertNull(params.get("a"));
        assertNull(params.get("b"));
    }

    @Test
    public void splitQueryNull() throws Exception {
        Map<String, String> params = NeatoAuthenticationResponse.fromUri(null).splitQuery(null);
        assertNotNull(params);
        assertNull(params.get("a"));
        assertNull(params.get("b"));
    }

    @Test
    public void getTokenExpirationDateNoChange() throws Exception {
        String inDateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'";
        SimpleDateFormat inFormat = new SimpleDateFormat(inDateFormat, Locale.US);

        Date date1 = inFormat.parse("2014-01-02T12:00:00Z");
        Date expire = NeatoAuthenticationResponse.fromUri(null).getTokenExpirationDate(date1, 0);//0 sec

        Calendar cal1 = Calendar.getInstance();
        cal1.setTime(expire);

        Calendar cal2 = Calendar.getInstance();
        cal2.setTime(date1);
        assertFalse(cal1.before(cal2));
        assertFalse(cal2.after(cal1));
    }

    @Test
    public void getTokenExpirationDate1Minute() throws Exception {
        String inDateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'";
        SimpleDateFormat inFormat = new SimpleDateFormat(inDateFormat, Locale.US);

        Date date1 = inFormat.parse("2014-01-02T12:00:00Z");
        Date expire = NeatoAuthenticationResponse.fromUri(null).getTokenExpirationDate(date1, 60);//60 sec
        Date date2 = inFormat.parse("2014-01-02T12:01:00Z");//add exactly 1 minute

        Calendar cal1 = Calendar.getInstance();
        cal1.setTime(expire);

        Calendar cal2 = Calendar.getInstance();
        cal2.setTime(date2);
        assertFalse(cal1.before(cal2));
        assertFalse(cal2.after(cal1));
    }

    @Test
    public void getTokenExpirationDate1Second() throws Exception {
        String inDateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'";
        SimpleDateFormat inFormat = new SimpleDateFormat(inDateFormat, Locale.US);

        Date date1 = inFormat.parse("2014-01-02T12:00:00Z");
        Date expire = NeatoAuthenticationResponse.fromUri(null).getTokenExpirationDate(date1, 1);//1 sec
        Date date2 = inFormat.parse("2014-01-02T12:00:01Z");//add exactly 1 sec

        Calendar cal1 = Calendar.getInstance();
        cal1.setTime(expire);

        Calendar cal2 = Calendar.getInstance();
        cal2.setTime(date2);
        assertFalse(cal1.before(cal2));
        assertFalse(cal2.after(cal1));
    }
}

File: Neato-SDK/neato-sdk-android/src/test/java/com/neatorobotics/sdk/android/authentication/NeatoAuthenticationResponseTest.kt
package com.neatorobotics.sdk.android.authentication

import android.net.Uri
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.MockitoAnnotations
import org.mockito.junit.MockitoJUnitRunner

import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.assertFalse
import org.mockito.Mockito.*

@RunWith(MockitoJUnitRunner::class)
class NeatoAuthenticationResponseTest {

    @Mock
    internal var uriToken: Uri? = null
    @Mock
    internal var uriError: Uri? = null
    @Mock
    internal var uriInvalid: Uri? = null

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)
        `when`(uriToken!!.toString()).thenReturn("my-app://neato#access_token=123&expiration=1")
        `when`(uriError!!.toString()).thenReturn("my-app://neato#error=error_1&error_description=description1")
        `when`(uriInvalid!!.toString()).thenReturn("")
    }

    @Test
    fun fromUriToken() {
        val response = NeatoAuthenticationResponse.fromUri(uriToken!!)
        assertNotNull(response)
        assertTrue(response.type === NeatoAuthenticationResponse.Response.TOKEN)
        assertEquals("123", response.token)
        assertNull(response.error)
        assertNull(response.errorDescription)
    }

    @Test
    fun fromUriError() {
        val response = NeatoAuthenticationResponse.fromUri(uriError!!)
        assertNotNull(response)
        assertTrue(response.type === NeatoAuthenticationResponse.Response.ERROR)
        assertEquals("error_1", response.error)
        assertEquals("description1", response.errorDescription)
        assertNull(response.token)
    }

    @Test
    fun fromUriCancelled() {
        val response = NeatoAuthenticationResponse.fromUri(uriInvalid!!)
        assertNotNull(response)
        assertTrue(response.type === NeatoAuthenticationResponse.Response.INVALID)
        assertNull(response.error)
        assertNull(response.errorDescription)
        assertNull(response.token)
    }

    @Test
    fun splitQueryMultipleParams() {
        val params = NeatoAuthenticationResponse.fromUri(null).splitQuery("a=vala&b=valb")
        assertNotNull(params)
        assertNotNull(params["a"])
        assertNotNull(params["b"])
        assertEquals(params["a"], "vala")
        assertEquals(params["b"], "valb")
    }

    @Test
    fun splitQuerySingleParam() {
        val params = NeatoAuthenticationResponse.fromUri(null).splitQuery("a=vala")
        assertNotNull(params)
        assertNotNull(params["a"])
        assertNull(params["b"])
        assertEquals(params["a"], "vala")
    }

    @Test
    fun splitQueryNoParam() {
        val params = NeatoAuthenticationResponse.fromUri(null).splitQuery("")
        assertNotNull(params)
        assertNull(params["a"])
        assertNull(params["b"])
    }

    @Test
    fun splitQueryNull() {
        val params = NeatoAuthenticationResponse.fromUri(null).splitQuery(null)
        assertNotNull(params)
        assertNull(params["a"])
        assertNull(params["b"])
    }

    @Test
    fun getTokenExpirationDateNoChange() {
        val inDateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        val inFormat = SimpleDateFormat(inDateFormat, Locale.US)

        val date1 = inFormat.parse("2014-01-02T12:00:00Z")
        val expire = NeatoAuthenticationResponse.fromUri(null).getTokenExpirationDate(date1, 0)//0 sec

        val cal1 = Calendar.getInstance()
        cal1.time = expire

        val cal2 = Calendar.getInstance()
        cal2.time = date1
        assertFalse(cal1.before(cal2))
        assertFalse(cal2.after(cal1))
    }

    @Test
    fun getTokenExpirationDate1Minute() {
        val inDateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        val inFormat = SimpleDateFormat(inDateFormat, Locale.US)

        val date1 = inFormat.parse("2014-01-02T12:00:00Z")
        val expire = NeatoAuthenticationResponse.fromUri(null).getTokenExpirationDate(date1, 60)//60 sec
        val date2 = inFormat.parse("2014-01-02T12:01:00Z")//add exactly 1 minute

        val cal1 = Calendar.getInstance()
        cal1.time = expire

        val cal2 = Calendar.getInstance()
        cal2.time = date2
        assertFalse(cal1.before(cal2))
        assertFalse(cal2.after(cal1))
    }

    @Test
    fun getTokenExpirationDate1Second() {
        val inDateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        val inFormat = SimpleDateFormat(inDateFormat, Locale.US)

        val date1 = inFormat.parse("2014-01-02T12:00:00Z")
        val expire = NeatoAuthenticationResponse.fromUri(null).getTokenExpirationDate(date1, 1)//1 sec
        val date2 = inFormat.parse("2014-01-02T12:00:01Z")//add exactly 1 sec

        val cal1 = Calendar.getInstance()
        cal1.time = expire

        val cal2 = Calendar.getInstance()
        cal2.time = date2
        assertFalse(cal1.before(cal2))
        assertFalse(cal2.after(cal1))
    }
}